File: solution2.py
import os
import time
import multiprocessing


def find_new_line(chunks: list[list[int]], file_with_ip: str) -> list[list[int]]:
    # go through each pair of [start, end]
    # adjust end to the closest start of the line
    # shift current end and next start

    with open(file_with_ip, 'br') as file:
        for i in range(len(chunks) - 1):
            shift = 0
            file.seek(chunks[i][1])
            while file.read(1) != b'\n':
                shift += 1
            chunks[i][1] += shift + 1
            chunks[i + 1][0] += shift + 1
    return chunks


def split_file_to_chunks(file_with_ip: str, process_count: int) -> list[list[int]]:
    # check filesize and create several ranges
    # to read file in chunks

    filesize = os.path.getsize(file_with_ip)
    chunks_len = filesize // process_count
    chunks = [[i, i + chunks_len] for i in range(0, filesize - process_count, chunks_len)]
    chunks[-1][1] = filesize
    chunks = find_new_line(chunks, file_with_ip)
    return chunks


def fill_bitmap(file_with_ip: str, bitmap_file: str, chunk: list[int, int]) -> None:
    start_pos = chunk[0]
    end_pos = chunk[1]
    with open(bitmap_file, 'rb+') as bitmap:
        with open(file_with_ip, 'r') as file:
            # move pointer to starting position
            file.seek(start_pos)
            while start_pos < end_pos:
                # read line by line in current chunk
                curr_line = file.readline()
                start_pos += len(curr_line)
                # convert ip to integer
                octets = curr_line.strip().split(".")
                ip_to_number = (int(octets[0]) << 24) | (int(octets[1]) << 16) | (int(octets[2]) << 8) | (
                    int(octets[3]))
                # set corresponding bit in the bitmap
                # and write it to the bitmap file
                bitmap.seek(ip_to_number // 8)
                byte = ord(bitmap.read(1))
                byte |= (1 << (7 - (ip_to_number % 8)))
                bitmap.seek(-1, 1)
                bitmap.write(bytes([byte]))


def count_unique_ips(bitmap_file: str, chunk: list[int, int]) -> int:
    # adjust ranges
    # and count set bits in the bytes of the bitmap file
    start_pos = chunk[0] // 8
    end_pos = chunk[1] // 8
    with open(bitmap_file, 'rb') as file:
        sm = 0
        file.seek(start_pos)
        while start_pos < end_pos:
            byte = file.read(1)
            sm += bin(ord(byte)).count('1')
            start_pos += 1
    return sm


def main() -> int:
    start_time = time.perf_counter()
    # little config
    # file_with_ip = "ip_addresses"
    # file_with_ip = 'test_data3.txt' # 1000
    # file_with_ip = "test_data.txt" # 9988184
    file_with_ip = "test_data2.txt"  # 98845647
    bitmap_file = "bitmap.dat"
    process_count = multiprocessing.cpu_count() // 2
    file_names = [file_with_ip] * process_count
    bitmap_names = [bitmap_file] * process_count
    with open(bitmap_file, 'wb') as file:
        file.write(bytearray(2 ** 29))

    # split file into chunks [a,b] to read in separate processes
    # then modify the ranges, so that a is always at the beginning of a new line
    chunks_ranges = split_file_to_chunks(file_with_ip, process_count)

    # read file and fill bitmap to count unique ip
    with multiprocessing.Pool(processes=process_count) as pool:
        pool.starmap(fill_bitmap, zip(file_names, bitmap_names, chunks_ranges))
    end_time = time.perf_counter()

    print("done reading file and filling bitmap, took: ", end_time - start_time)

    start_time = time.perf_counter()

    # create chunks of [a,b] to split entire bitmap
    chunks_len = 2 ** 32 // process_count
    chunks_ranges = [[i, i + chunks_len] for i in range(0, 2 ** 32 - process_count, chunks_len)]
    chunks_ranges[-1][1] = 2 ** 32

    with multiprocessing.Pool(processes=process_count) as pool:
        chunk_sum = pool.starmap(count_unique_ips, zip(bitmap_names, chunks_ranges))
    res = sum(chunk_sum)

    end_time = time.perf_counter()
    print("done counting unique number, took: ", end_time - start_time)
    # remove file if needed
    # os.remove(bitmap_file)
    return res


if __name__ == "__main__":
    result = main()
    print(result)

File: solution2.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"math/bits"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// go through each pair of [start, end]
// adjust end to the closest start of the line
// shift current end and next start
func findNewLine(chunks [][]int64, fileWithIP string) {
	file, err := os.Open(fileWithIP)
	if err != nil {
		panic(err)
	}
	defer file.Close()
	reader := bufio.NewReader(file)

	for i := 0; i < len(chunks)-1; i++ {
		file.Seek(int64(chunks[i][1]), io.SeekStart)
		var line string
		line, _ = reader.ReadString('\n')
		reader.Discard(reader.Buffered())
		shift := int64(len(line))
		chunks[i][1] += shift
		chunks[i+1][0] += shift
	}
}

// create pairs of chunks depending on the filesize
func splitFileToChunks(fileWithIP string, processCount int) [][]int64 {
	fileInfo, err := os.Stat(fileWithIP)
	if err != nil {
		panic(err)
	}
	filesize := fileInfo.Size()

	chunksLen := filesize / int64(processCount)
	chunks := make([][]int64, processCount)
	for i := 0; i < processCount-1; i++ {
		chunks[i] = []int64{int64(i) * chunksLen, int64(i+1) * chunksLen}
	}
	chunks[processCount-1] = []int64{int64(processCount-1) * chunksLen, filesize}
	findNewLine(chunks, fileWithIP)
	return chunks
}

// read each IP
// convert it into int
// set the corrsponding bit to 1
func fillBitmap(fileWithIPName string, bitMapFileName string, chunk []int64) { //  wg *sync.WaitGroup
	var start int64 = chunk[0]
	var end int64 = chunk[1]
	ipFile, _ := os.Open(fileWithIPName)
	defer ipFile.Close()
	bitmapFile, _ := os.OpenFile(bitMapFileName, os.O_RDWR, 0644)
	defer bitmapFile.Close()

	ipFile.Seek(int64(chunk[0]), io.SeekStart)
	reader := bufio.NewReader(ipFile)
	for start < end {
		ip, err := reader.ReadString('\n')
		if err == io.EOF {
			break
		}
		start += int64(len(ip))
		ip = strings.TrimSpace(ip)
		parts := strings.Split(string(ip), ".")
		var octets [4]int
		for i := 0; i < 4; i++ {
			octets[i], err = strconv.Atoi(parts[i])
		}
		convertedIP := uint32(octets[0]<<24 | octets[1]<<16 | octets[2]<<8 | octets[3])
		bytePos := int64(convertedIP / 8)
		bitOffset := convertedIP % 8

		oneByte := make([]byte, 1)
		_, err = bitmapFile.ReadAt(oneByte, bytePos)

		oneByte[0] |= (1 << (7 - bitOffset))
		_, err = bitmapFile.WriteAt(oneByte, bytePos)

	}

}

func countUniqueIPs(bitmapFile string, chunk []int64) uint32 {
	startPos := chunk[0] / 8
	endPos := chunk[1] / 8

	file, _ := os.Open(bitmapFile)

	file.Seek(startPos, io.SeekStart)

	var sum uint32
	reader := bufio.NewReader(file)

	for startPos < endPos {
		oneByte, _ := reader.ReadByte()
		startPos += 1
		sum += uint32(bits.OnesCount8(oneByte))
	}
	return sum
}

// create pairs to go through each of 2^32 nums
// to check later for set bits
func splitRangeToChunks(processCount int) [][]int64 {
	chunks := make([][]int64, processCount)
	maxLen := int64(1 << 32)
	chunksLen := maxLen / int64(processCount)
	for i := 0; i < processCount-1; i++ {
		chunks[i] = []int64{int64(i) * chunksLen, int64(i+1) * chunksLen}
	}
	chunks[processCount-1] = []int64{int64(processCount-1) * chunksLen, maxLen + 1}

	return chunks
}

func main() {
	// config
	// fileWithIP := "ip_addresses"
	fileWithIP := "test_data3.txt"
	bitmapFileName := "bitmap.dat"
	processCount := 10

	bitmapFile, _ := os.Create(bitmapFileName)
	defer bitmapFile.Close()
	bitmapFile.Truncate((1 << 29))

	chunks := splitFileToChunks(fileWithIP, processCount)

	start := time.Now()
	var wg sync.WaitGroup
	wg.Add(processCount)
	for i := 0; i < processCount; i++ {
		go func(i int) {
			defer wg.Done()
			fillBitmap(fileWithIP, bitmapFileName, chunks[i])
		}(i)
	}

	wg.Wait()
	end := time.Now()
	fmt.Println("done reading file and filling bitmap, took: ", end.Sub(start))

	rangeChunks := splitRangeToChunks(processCount)
	var result uint32
	wg.Add(processCount)
	for i := 0; i < processCount; i++ {
		go func(i int) {
			defer wg.Done()
			result += countUniqueIPs(bitmapFileName, rangeChunks[i])

		}(i)
	}
	wg.Wait()
	fmt.Println(result)

}
